File: lib/tools/meta-ads.ts
import { getCached, setCache } from '../supabase';
import { scrapePage } from './firecrawl';
import type { ToolResult, MetaAd } from './types';
import { buildToolResult } from './fallback';

// ── Meta Ad Library — browser scrape via Firecrawl ───────────────────────────
// The official API only covers political/EU ads.
// The public Ad Library page covers all advertisers with no token required.

const AD_LIBRARY_BASE = 'https://www.facebook.com/ads/library';

function buildAdLibraryUrl(advertiserName: string): string {
  const params = new URLSearchParams({
    active_status: 'all',
    ad_type: 'all',
    country: 'US',
    q: advertiserName,
    search_type: 'keyword_unordered',
  });
  return `${AD_LIBRARY_BASE}/?${params}`;
}

// Extract ad copy snippets from scraped markdown
function parseAdsFromMarkdown(markdown: string, advertiserName: string): MetaAd[] {
  const ads: MetaAd[] = [];

  // Split on common ad boundary patterns in the scraped content
  const blocks = markdown
    .split(/\n{2,}/)
    .map(b => b.trim())
    .filter(b => b.length > 30 && b.length < 1000);

  blocks.slice(0, 15).forEach((block, i) => {
    // Skip nav/footer noise
    if (/cookie|privacy|terms|sign in|log in/i.test(block)) return;

    ads.push({
      id: `scraped-${i}`,
      page_name: advertiserName,
      ad_creative_body: block,
      ad_snapshot_url: buildAdLibraryUrl(advertiserName),
    });
  });

  return ads.slice(0, 8);
}

export async function searchMetaAds(
  advertiserName: string,
  _country = 'US',
  _limit = 15,
): Promise<ToolResult<MetaAd[]>> {
  const cacheKey = `meta:browser:${advertiserName}`;
  const cached = await getCached('meta_ads', cacheKey);
  if (cached) return { ...(cached as ToolResult<MetaAd[]>), cached: true };

  const url = buildAdLibraryUrl(advertiserName);

  try {
    const scraped = await scrapePage(url);
    const ads = parseAdsFromMarkdown(scraped.data.markdown, advertiserName);

    const result = buildToolResult<MetaAd[]>({
      data: ads,
      status: ads.length > 0 ? 'degraded' : 'failed',
      source: 'Meta Ad Library (browser scrape)',
      sourceUrl: url,
    });

    await setCache('meta_ads', cacheKey, result);
    return result;
  } catch {
    return buildToolResult<MetaAd[]>({
      data: [],
      status: 'failed',
      source: 'Meta Ad Library (browser scrape)',
      sourceUrl: url,
    });
  }
}

export async function getAdMessaging(advertiserName: string): Promise<string[]> {
  const result = await searchMetaAds(advertiserName);
  return result.data
    .map(ad => ad.ad_creative_body ?? ad.ad_creative_link_title ?? '')
    .filter(Boolean)
    .slice(0, 5);
}

File: backend/app/tools/meta_ads.py
"""Meta Ad Library browser scrape — mirrors lib/tools/meta-ads.ts."""

from __future__ import annotations

import re
from typing import Any
from urllib.parse import urlencode

from app.supabase_client import get_cached, set_cache

from .fallback import build_tool_result
from .firecrawl import scrape_page

AD_LIBRARY_BASE = "https://www.facebook.com/ads/library"


def _build_ad_library_url(advertiser_name: str) -> str:
    params = urlencode(
        {
            "active_status": "all",
            "ad_type": "all",
            "country": "US",
            "q": advertiser_name,
            "search_type": "keyword_unordered",
        }
    )
    return f"{AD_LIBRARY_BASE}/?{params}"


def _parse_ads_from_markdown(markdown: str, advertiser_name: str) -> list[dict[str, Any]]:
    ads: list[dict[str, Any]] = []
    blocks = [
        b.strip()
        for b in re.split(r"\n{2,}", markdown)
        if 30 < len(b.strip()) < 1000
    ]

    for i, block in enumerate(blocks[:15]):
        if re.search(r"cookie|privacy|terms|sign in|log in", block, re.I):
            continue
        ads.append(
            {
                "id": f"scraped-{i}",
                "page_name": advertiser_name,
                "ad_creative_body": block,
                "ad_snapshot_url": _build_ad_library_url(advertiser_name),
            }
        )

    return ads[:8]


async def search_meta_ads(
    advertiser_name: str,
    country: str = "US",
    limit: int = 15,
) -> dict[str, Any]:
    del country, limit  # API parity with TS; scrape uses US keyword search

    cache_key = f"meta:browser:{advertiser_name}"
    cached = await get_cached("meta_ads", cache_key)
    if cached:
        return {**cached, "cached": True}

    url = _build_ad_library_url(advertiser_name)

    try:
        scraped = await scrape_page(url)
        ads = _parse_ads_from_markdown(
            (scraped.get("data") or {}).get("markdown") or "",
            advertiser_name,
        )
        result = build_tool_result(
            data=ads,
            status="degraded" if ads else "failed",
            source="Meta Ad Library (browser scrape)",
            source_url=url,
        )
        await set_cache("meta_ads", cache_key, result)
        return result
    except Exception:
        return build_tool_result(
            data=[],
            status="failed",
            source="Meta Ad Library (browser scrape)",
            source_url=url,
        )


async def get_ad_messaging(advertiser_name: str) -> list[str]:
    result = await search_meta_ads(advertiser_name)
    messages: list[str] = []
    for ad in result.get("data") or []:
        body = ad.get("ad_creative_body") or ad.get("ad_creative_link_title") or ""
        if body:
            messages.append(body)
        if len(messages) >= 5:
            break
    return messages
